src/builder-source-inline.c

```diff
@@ -0,0 +1,233 @@
+#include "config.h"
+
+#include <string.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/statfs.h>
+
+#include "builder-flatpak-utils.h"
+#include "builder-utils.h"
+#include "builder-source-inline.h"
+
+struct BuilderSourceInline
+{
+  BuilderSource parent;
+
+  char         *contents;
+  gboolean      base64;
+  char         *dest_filename;
+};
+
+typedef struct
+{
+  BuilderSourceClass parent_class;
+} BuilderSourceInlineClass;
+
+G_DEFINE_TYPE (BuilderSourceInline, builder_source_inline, BUILDER_TYPE_SOURCE);
+
+enum {
+  PROP_0,
+  PROP_CONTENTS,
+  PROP_BASE64,
+  PROP_DEST_FILENAME,
+  LAST_PROP
+};
+
+static void
+builder_source_inline_finalize (GObject *object)
+{
+  BuilderSourceInline *self = (BuilderSourceInline *) object;
+
+  g_free (self->contents);
+  g_free (self->dest_filename);
+
+  G_OBJECT_CLASS (builder_source_inline_parent_class)->finalize (object);
+}
+
+static void
+builder_source_inline_get_property (GObject    *object,
+                                    guint       prop_id,
+                                    GValue     *value,
+                                    GParamSpec *pspec)
+{
+  BuilderSourceInline *self = BUILDER_SOURCE_INLINE (object);
+
+  switch (prop_id)
+    {
+    case PROP_CONTENTS:
+      g_value_set_string (value, self->contents);
+      break;
+
+    case PROP_BASE64:
+      g_value_set_boolean(value, self->base64);
+      break;
+
+    case PROP_DEST_FILENAME:
+      g_value_set_string (value, self->dest_filename);
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+static void
+builder_source_inline_set_property (GObject      *object,
+                                    guint         prop_id,
+                                    const GValue *value,
+                                    GParamSpec   *pspec)
+{
+  BuilderSourceInline *self = BUILDER_SOURCE_INLINE (object);
+
+  switch (prop_id)
+    {
+    case PROP_CONTENTS:
+      g_free (self->contents);
+      self->contents = g_value_dup_string (value);
+      break;
+
+    case PROP_BASE64:
+      self->base64 = g_value_get_boolean (value);
+      break;
+
+    case PROP_DEST_FILENAME:
+      g_free (self->dest_filename);
+      self->dest_filename = g_value_dup_string (value);
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+static gboolean
+builder_source_inline_validate (BuilderSource  *source,
+                                GError        **error)
+{
+  BuilderSourceInline *self = BUILDER_SOURCE_INLINE (source);
+
+  if (self->dest_filename != NULL &&
+      strchr (self->dest_filename, '/') != NULL)
+    return flatpak_fail (error, "No slashes allowed in dest-filename");
```

Comment:
```suggestion
    return flatpak_fail (error, "No slashes allowed in dest-filename, use dest property for directory");
```

Reply:
This was copied over from script source. Maybe we should change this message for all the other source types?

Reply:
I think so, its just helpful.

Reply:
Done.